python/snewpy/flavor.py

```diff
@@ -0,0 +1,181 @@
+        
+    def __repr__(self):
+        s=self._repr_short()+'\n'+repr(self.array)
+        return s
+    def __eq__(self,other):
+        return self.flavor_in==other.flavor_in and self.flavor_out==other.flavor_out and np.allclose(self.array,other.array)
+                
+    def __matmul__(self, other):
+        if isinstance(other, FlavorMatrix):
+            try:
+                data = np.tensordot(self.array, other.array, axes=[1,0])
+                return FlavorMatrix(data, self.flavor_out, from_flavor = other.flavor_in)
+            except Exception as e:
+                raise ValueError(f"Cannot multiply {self._repr_short()} by {other._repr_short()}") from e
+        elif hasattr(other, '__rmatmul__'):
+            return other.__rmatmul__(self)        
+        raise TypeError(f"Cannot multiply object of {self.__class__} by {other.__class__}")
+    #properties
+    @property
+    def shape(self):
+        return self.array.shape
+    @property
+    def flavor(self):
+        return self.flavor_out
+        
+    @classmethod
+    def zeros(cls, flavor:FlavorScheme, from_flavor:FlavorScheme = None):
+        from_flavor = from_flavor or flavor
+        shape = (len(from_flavor), len(flavor))
+        data = np.zeros(shape)
+        return cls(data, flavor, from_flavor)
+        
+    @classmethod
+    def eye(cls, flavor:FlavorScheme, from_flavor:FlavorScheme = None):
+        from_flavor = from_flavor or flavor
+        shape = (len(from_flavor), len(flavor))
+        data = np.eye(*shape)
+        return cls(data, flavor, from_flavor)
+
+    @classmethod
+    def from_function(cls, flavor:FlavorScheme, from_flavor:FlavorScheme = None):
+        """A decorator for creating the flavor matrix from the given function"""
+        from_flavor = from_flavor or flavor
+        def _decorator(function):
+            data = [[function(f1,f2)
+                     for f2 in from_flavor]
+                    for f1 in flavor]
+                    
+            return cls(np.array(data,dtype=float),  flavor, from_flavor)
+        return _decorator
+    #flavor conversion utils
+    
+def conversion_matrix(from_flavor:FlavorScheme, to_flavor:FlavorScheme):
+    if(from_flavor==TwoFlavor):
+        #define special cases
+        @FlavorMatrix.from_function(to_flavor, from_flavor)
+        def convert_2toN(f1,f2):
+            if (f1.name==f2.name):
+                return 1.
+            if (f1.is_neutrino==f2.is_neutrino)and(f2.lepton=='X' and f1.lepton in ['MU','TAU']):
+                return 1.
+            return 0
+        return convert_2toN
+    else:
+        @FlavorMatrix.from_function(to_flavor, from_flavor)
+        def convert_Nto2(f1,f2):
+            if (f1.name==f2.name):
+                return 1.
+            if (f1.is_neutrino==f2.is_neutrino)and(f1.lepton=='X' and f2.lepton in ['MU','TAU']):
+                return 0.5
+            return 0.
+        return convert_Nto2
```

Comment:
You're right, and your version is more compact.

About losing the information - you're also correct, but this can be exactly what user wants.For example to get rid of steriles.

So maybe we could issue a warning when losing information, instead of just refusing?

On the other hand, in most cases this conversion is done explicitly by user - by calling conversion operators or multiplying by the matrix. IMO we don't have to predict and prevent the user's incorrect actions - otherwise we should also check if flux is multiplied by 0, and also loses information 😃 

Maybe this is a question for discussion in #309: what is the user interface and restrictions.

Reply:
I think there's a degree of uncertainty here; since applying a particular matrix to one flux might be totally fine, but applying it to a different flux could lose information.

You're right, explicitly dropping steriles is a use case we should support; so a warning instead of an error would be better.